Support DSR with Geneve dispatch

ysksuzuki opened the pull request:
For reviewers,

1. Currently, this implementation doesn't support IPv6 with XDP. It works with IPv4 either in XDP or TC, relying on XDP to TC punt. To support IPv6 in XDP, we need to implement the XDP native encap, which planning to do as a follow-up. 
2. This implementation is compatible with the Geneve tunneling mode, which means that it works with either the direct routing mode or the Geneve tunneling mode. but unfortunately, it doesn't work with VXlan tunnel because there's no space to store service IP/port in the VXlan header. 
3. ~Picked up an unassigned Geneve option class and type, referencing~ We have been allocated 0x014B https://www.iana.org/assignments/nvo3/nvo3.xhtml#geneve-option-class. 
4. ~Trying to mitigate the MTU overhead by encapsulating only SYN packets as the IP option mode does~ Encapsulate the whole connection, but only add the GENEVE option to the TCP-SYN packets


Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->

Fixes: #22955

```release-note
Support DSR with Geneve dispatch in CNI mode
```

ysksuzuki committed:
test: add test for DSR with Geneve

Signed-off-by: Yusuke Suzuki <yusuke-suzuki@cybozu.co.jp>

ysksuzuki committed:
docs: Add description for DSR with Geneve

Signed-off-by: Yusuke Suzuki <yusuke-suzuki@cybozu.co.jp>

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1597,14 +1679,66 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
+		__u32 addr = ctx_load_meta(ctx, CB_ADDR_V4);
+
+		set_geneve_opt4(port, addr, &gopt);
+
```
Note that the current IPv4 DSR code is confused at this point. The addr/port is actually in big-endian format - and we should preserve that in our Geneve DSR Option, all the way to the backend without any conversion.

Also `tail_nodeport_ipv4_dsr()` has already loaded them from the CB, so we can just pass them in as parameters.

ysksuzuki replied:
That is what you mentioned in https://github.com/cilium/cilium/pull/23262, isn't it? I will fix it.

julianwiedmann replied:
> That is what you mentioned in #23262, isn't it? I will fix it.

yes, exactly. Let's not repeat that mistake :)

julianwiedmann replied:
Looks much more straight-forward now :)

julianwiedmann commented on bpf/lib/common.h:
```diff
@@ -1095,6 +1112,28 @@ struct lpm_val {
+	__u8	r2:1;
+	__u8	r1:1;
+	__u32	opt_data[2];
+};
```
Let's spell this out as
```
__be32 addr;
__be16 port;
__u16 pad;
```

(same for the IPv6 option)

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1626,6 +1760,19 @@ nodeport_extract_dsr_v4(struct __ctx_buff *ctx, struct iphdr *ip4,
+			return 0;
+		}
+	}
+#else
```
This will only be utilized in `from-overlay`, correct? Can we add a `defined(IS_BPF_OVERLAY)` condition? That should also allow us to remove the XDP stub for `ctx_get_tunnel_opt()`.

julianwiedmann commented on bpf/bpf_overlay.c:
```diff
@@ -305,8 +312,19 @@ static __always_inline int handle_ipv4(struct __ctx_buff *ctx, __u32 *identity)
 					   METRIC_INGRESS, false, false);
+#endif
 	}
 
```
I have a feeling we should/need to integrate this into `ipv4_local_delivery()` better - passing a `deliver-via-kernel` flag or such.

Also need to be careful about interactions with Egress-Gateway reply traffic when in Direct-Routing mode ...

ysksuzuki replied:
> Also need to be careful about interactions with Egress-Gateway reply traffic when in Direct-Routing mode ...

Ah, right, I will check the interaction with Egress GW

julianwiedmann replied:
I'm still a bit stuck on this part - primarily because it changes behaviour for EgressGW reply traffic, as it passes through `from-overlay` on the worker node. Can you sketch out a bit more why the change is needed, and how come we didn't need it for EgressGW traffic before?

ysksuzuki replied:
This change is intended primarily for the generic veth chain use case. In the chaining mode, the other chained plugin is responsible for the underlying connectivity. I thought that bypassing the kernel stack here would break this assumption.
https://github.com/cilium/cilium/blob/faa23a078767f1892325f15fceaa7b2fc14f3ca8/install/kubernetes/cilium/templates/cilium-configmap.yaml#L309-L313

EgressGW also supports the generic veth chain mode, right? If so, shouldn't we do this in the first place?

ysksuzuki replied:
As far as I looked into the packet flow under the CNI chain mode, we pass packets to the kernel stack here and it delivers them to the local endpoint. I think we should do this on the bpf_overlay as well
https://github.com/cilium/cilium/blob/d56653d6f63557088c4c5065ff9cf2796777df9a/bpf/bpf_host.c#L534

ysksuzuki replied:
Dropped the deliver via kernel commit for now. Let's discuss it as a follow-on

julianwiedmann replied:
Ack, thanks!

julianwiedmann reviewed (commented):
Hi @ysksuzuki - just a few comments from a quick look. Haven't poked at the XDP-to_TC logic yet. Looking really good so far! Please consider if you can split the patches up somewhat - maybe one for the DSR insertion, one for the DSR extraction, one for the agent ...?

My initial expectation was that we would encapsulate the *whole* DSR connection in Geneve, not just the TCP-SYN. Seeing this mixed pattern in network traces will be fun when debugging :).

ysksuzuki commented on bpf/bpf_host.c:
```diff
@@ -1039,11 +1039,28 @@ int cil_from_netdev(struct __ctx_buff *ctx)
+
+			set_geneve_opt4(port, addr, &gopt);
+
+			return encap_and_redirect_with_nodeid_opt4(ctx,
```
This code doesn't work when handling the reply packets from the Egress GW. It should call `__encap_and_redirect_with_nodeid` instead in that case. Will fix it.

ysksuzuki committed:
bpf: nodeport: Encapsulate DSR flow with Geneve

This commit encapsulates the DSR flow with Geneve, inserts DSR specific
option, the service IP/port, into the Geneve option and redirects to
the selected backend. Currently, it doesn't support IPv6 with XDP.

Signed-off-by: Yusuke Suzuki <yusuke-suzuki@cybozu.co.jp>

ysksuzuki committed:
bpf: nodeport: Extract DSR specific option from the Geneve header

Let the dsr code extract the DSR specific option, the service IP/port,
from the Geneve header and integrate with the DSR implementation on the
receiver side.

Signed-off-by: Yusuke Suzuki <yusuke-suzuki@cybozu.co.jp>

ysksuzuki committed:
daemon: Support DSR with Geneve dispatch

This commit adds DSR with Geneve dispatch mode. Geneve dispatch mode is
compatible with the Geneve tunneling mode, which means that it works with
either the direct routing mode or the Geneve tunneling mode. but unfortunately,
it doesn't work with vxlan tunnel because there's no space to store service
IP/port in the vxlan header.

Signed-off-by: Yusuke Suzuki <yusuke-suzuki@cybozu.co.jp>

julianwiedmann commented on bpf/lib/common.h:
```diff
@@ -1106,6 +1123,35 @@ struct lpm_val {
+	__be32	addr1;
+	__be32	addr2;
+	__be32	addr3;
+	__be32	addr4;
```
Have you tried turning this into a `union v6addr`, same as for the IPv6 Extension header? Then we can also use `ipv6_addr_copy()`.

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1438,6 +1525,25 @@ nodeport_extract_dsr_v4(struct __ctx_buff *ctx, struct iphdr *ip4,
+			}
+		}
+	}
+#else
```
With this we would still walk through the TCP-specific code above and below, right? So I would take this even further - only call `nodeport_extract_dsr_v4()` iff

1. `DSR_ENCAP_MODE == GENEVE` && `IS_BPF_OVERLAY`, or
2. `DSR_ENCAP_MODE != GENEVE` && !`IS_BPF_OVERLAY`

And then have a simpler condition in here - if `defined(IS_BPF_OVERLAY)` look for the tunnel-opt, else look for IP-Option.

julianwiedmann commented on bpf/lib/encap.h:
```diff
@@ -295,5 +295,98 @@ encap_and_redirect_netdev(struct __ctx_buff *ctx, struct tunnel_key *k,
+	gopt->type = DSR_GENEVE_OPT_TYPE;
+	gopt->r1 = 0;
+	gopt->r2 = 0;
+	gopt->r3 = 0;
```
No need to clear these explicitly, we have a full memset already.

julianwiedmann reviewed (changes requested):
Thanks @ysksuzuki ! Just a few smaller things, still need to digest the rest. So maybe hold off for a bit ...

Do we need to reduce the MTU accordingly to the max GENEVE option-size? :thinking:

julianwiedmann commented on bpf/lib/l3.h:
```diff
@@ -145,14 +150,19 @@ static __always_inline int ipv4_local_delivery(struct __ctx_buff *ctx, int l3_of
+		ctx_change_type(ctx, PACKET_HOST);
+		return CTX_ACT_OK;
+	}
+
```
To make review easier, could you please split the `deliver_via_kernel` change off into a separate commit, that only refactors the existing code (== introduces the `false` parameter for all current users) ? So the actual DSR patches don't touch any unrelated files.

julianwiedmann commented on bpf/lib/common.h:
```diff
@@ -1106,6 +1123,35 @@ struct lpm_val {
 };
 
+#if defined(ENABLE_DSR) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE
+struct geneve_opt4 {
```
Let's maybe already call this `geneve_dsr_opt4` - I'm expecting we will start to use other option types over time.

Btw, feel free to grab https://github.com/cilium/cilium/pull/24240/commits/541b2a740328b81a842a6d2c25fab1f090e223cb for a shared definition of the Options header :).

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1019,7 +1019,28 @@ int cil_from_netdev(struct __ctx_buff *ctx)
+		{
+			struct geneve_opt4 gopt;
+			__u16 port = (__u16)ctx_get_xfer(ctx, XFER_ENCAP_PORT);
+			__u32 addr = ctx_get_xfer(ctx, XFER_ENCAP_ADDR);
```
These are `__be16` / `__be32`, right? Just want to make sure that we have the endianess consistent everywhere.

ysksuzuki replied:
Yes, I will fix it

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -238,6 +238,34 @@ static __always_inline int dsr_set_ext6(struct __ctx_buff *ctx,
+						 ifindex);
+	}
+
+	return CTX_ACT_DROP;
```
We shouldn't return CTX_ACT_DROP as error condition from our own code. Can you make this a proper DROP reason?

ysksuzuki replied:
Sure, will do

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1403,10 +1465,35 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
+		set_geneve_opt4(svc_port, svc_addr, &gopt);
+
+		return  __encap_with_nodeid_opt4(ctx, info->tunnel_endpoint,
+						 WORLD_ID,
```
Ah, this is interesting. I can see how this would obviously work when in TUNNEL_MODE. But we also expect to run with direct-routing, correct? And only establish the tunnel-infrastructure for DSR, so that we can encapsulate. Would the `tunnel_endpoint` be populated in that case?

(we've only used this tunnel-with-direct-routing-mode for EgressGW so far, and there we get the tunnel_endpoint IP from the EgressGW policy entry ...)

ysksuzuki replied:
Yes, exactly. `tunnel_endpoint` will be populated with direct-routing mode as well, and it works

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1438,6 +1525,25 @@ nodeport_extract_dsr_v4(struct __ctx_buff *ctx, struct iphdr *ip4,
+		ret = ctx_get_tunnel_opt(ctx, &gopt, sizeof(gopt));
+
+		if (ret > 0) {
+			if (bpf_ntohs(gopt.opt_class) == DSR_GENEVE_OPT_CLASS &&
```
```suggestion
			if (gopt.opt_class == bpf_htons(DSR_GENEVE_OPT_CLASS) &&
```

Small trick that allows us to perform the byte-swap at compile-time.

julianwiedmann reviewed (changes requested):
Took a closer look - some smaller change requests.

As reminder to myself, there's three things that need an additional pass

1. the deliver-via-kernel parts
2. operation when in direct-routing mode
3. MTU (@brb suggested to have a look at `dsr_reply_icmp4()` instead)

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1403,10 +1465,35 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
+		set_geneve_opt4(svc_port, svc_addr, &gopt);
+
+		return  __encap_with_nodeid_opt4(ctx, info->tunnel_endpoint,
+						 WORLD_ID,
```
I overlooked this earlier - can we keep the TCP optimization? Sending the whole connection over the tunnel, but only adding the GENEVE option to the TCP-SYN.

ysksuzuki replied:
Sure, I will do that

ysksuzuki committed:
bpf: nodeport: Optimize the GENEVE encapsulation for the TCP flow

This commit optimizes the tunneling overhead by only adding the GENEVE
option to the TCP-SYN packets, sending the whole connection over
the tunnel

Signed-off-by: Yusuke Suzuki <yusuke-suzuki@cybozu.co.jp>

julianwiedmann commented on bpf/lib/encap.h:
```diff
@@ -295,5 +295,90 @@ encap_and_redirect_netdev(struct __ctx_buff *ctx, struct tunnel_key *k,
+		return ret_opt;
+
+	return ret;
+}
```
One thing I realized while considering in-XDP support - I believe that in the end, we want `ctx_set_encap_info()` to contain both the `ctx_set_tunnel_key()` and the `ctx_set_tunnel_opt()` operations. For TC it doesn't make any difference to keep them split up. But for XDP we need the total length of the encap headers (including the options), so that we can adjust the headroom in a single operation.

So essentially pipe additional `void *opt, __u32 opt_len` parameters through, and skip the `skb_set_tunnel_opt()` if `opt_len == 0`.

Just noting this here so we don't forget. Not sure if it would simplify your current implementation.

ysksuzuki replied:
Is it something like this? I think it is worth trying to simplify the current implementation

```c
static __always_inline int
__encap_with_nodeid(struct __ctx_buff *ctx, __be32 tunnel_endpoint,
		    __u32 seclabel, __u32 dstid, __u32 vni __maybe_unused,
		    enum trace_reason ct_reason, __u32 monitor, int *ifindex
		    void *opt, __u32 opt_len)
{
	__u32 node_id;
	int ret;

	/* When encapsulating, a packet originating from the local host is
	 * being considered as a packet from a remote node as it is being
	 * received.
	 */
	if (seclabel == HOST_ID)
		seclabel = LOCAL_NODE_ID;

	node_id = bpf_ntohl(tunnel_endpoint);

	cilium_dbg(ctx, DBG_ENCAP, node_id, seclabel);

	ret = ctx_set_encap_info(ctx, node_id, seclabel, dstid, vni, ifindex, opt, opt_len);
	if (ret == CTX_ACT_REDIRECT)
		send_trace_notify(ctx, TRACE_TO_OVERLAY, seclabel, dstid, 0, *ifindex,
				  ct_reason, monitor);

	return ret;
}

static __always_inline __maybe_unused int
ctx_set_encap_info(struct __sk_buff *ctx, __u32 node_id, __u32 seclabel,
		   __u32 dstid __maybe_unused, __u32 vni __maybe_unused,
		   int *ifindex, void *opt, __u32 opt_len)
{
	struct bpf_tunnel_key key = {};
	int ret;

#ifdef ENABLE_VTEP
	if (vni != NOT_VTEP_DST)
		key.tunnel_id = vni;
	else
#endif /* ENABLE_VTEP */
		key.tunnel_id = seclabel;

	key.remote_ipv4 = node_id;
	key.tunnel_ttl = IPDEFTTL;

	ret = ctx_set_tunnel_key(ctx, &key, sizeof(key), BPF_F_ZERO_CSUM_TX);
	if (unlikely(ret < 0))
		return DROP_WRITE_ERROR;

	if (opt_len > 0) {
		 int ret = ctx_set_tunnel_opt(ctx, opt, opt_len);
		if (unlikely(ret < 0))
			return DROP_WRITE_ERROR;
	}

	*ifindex = ENCAP_IFINDEX;

	return CTX_ACT_REDIRECT;
}
```

julianwiedmann replied:
>Is it something like this? I think it is worth trying to simplify the current implementation

yep that matches what I had in mind :+1:.

If it already makes your current implementation simpler, then let's go for it :).

ysksuzuki committed:
bpf: nodeport: Add the dsr_is_too_big check

This commit checks the expanded packet length when encapsulating the dsr
flow with GENEVE header and if it exceeds the device MTU, then returns
ICMP Destination Unreachable/Fragmentation needed.

Signed-off-by: Yusuke Suzuki <yusuke-suzuki@cybozu.co.jp>

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -238,6 +238,73 @@ static __always_inline int dsr_set_ext6(struct __ctx_buff *ctx,
+	union v6addr *dst;
+	bool need_opt = true;
+	__u16 encap_len = sizeof(struct ipv6hdr) + sizeof(struct udphdr) +
+		sizeof(struct genevehdr) + sizeof(struct geneve_dsr_opt6) + ETH_HLEN;
```
With TCP we should fine-tune this, and only consider the Option for SYN packets.

julianwiedmann commented on bpf/lib/overloadable_xdp.h:
```diff
@@ -167,12 +177,25 @@ ctx_set_encap_info(struct xdp_md *ctx __maybe_unused,
+		ctx_store_meta(ctx, CB_ENCAP_PORT, *port);
+		ctx_store_meta(ctx, CB_ENCAP_ADDR, *addr);
+	}
+#endif
```
Should we return a DROP reason if it's IPv6, just to be clear that it is unhandled currently?

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -482,10 +571,18 @@ int tail_nodeport_ipv6_dsr(struct __ctx_buff *ctx)
+			cilium_capture_out(ctx);
+			return ctx_redirect(ctx, ifindex, 0);
+		}
+	}
```
Maybe add a comment here why the code isn't looking the same as for IPv4 (the XDP-to-TC punt isn't needed).

julianwiedmann reviewed (commented):
Just a few minor things - think we're ready to switch this PR from draft to review, great work @ysksuzuki :)

kaworu reviewed (approved):
Hubble API changes LGTM.

kaworu commented on api/v1/flow/flow.proto:
```diff
@@ -375,6 +375,7 @@ enum DropReason {
     NAT46 = 187;
     NAT64 = 188;
     AUTH_REQUIRED = 189;
+    DSR_ENCAP_GENEVE_NOT_SUPPORT_IPV6_XDP = 190;
```
nitpicking: the name looks a bit long to me, how about
```suggestion
    DSR_GENEVE_IPV6_XDP_UNSUPPORTED = 190;
```

ysksuzuki replied:
Thank you for the review! I will make it shorter.

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1611,6 +1806,7 @@ int tail_nodeport_ipv4_dsr(struct __ctx_buff *ctx)
 	__s8 ext_err = 0;
 	__be32 addr;
 	__be16 port;
+	int ifindex __maybe_unused;
```
I think we can just use the `oif` instead.

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1403,10 +1512,77 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
+	if (ip4->protocol == IPPROTO_TCP) {
+		union tcp_flags tcp_flags = { .value = 0 };
+
+		if (ctx_load_bytes(ctx, ETH_HLEN + sizeof(*ip4) + 12,
```
```suggestion
		if (l4_load_tcp_flags(ctx, ETH_HLEN + ipv4_hdrlen(ip4),
```

julianwiedmann commented on bpf/bpf_overlay.c:
```diff
@@ -299,14 +305,25 @@ static __always_inline int handle_ipv4(struct __ctx_buff *ctx, __u32 *identity)
+		 * e.g. CNI chain setup
+		 */
+		deliver_via_kernel = true;
+#endif
```
In this case, where/when would we apply the endpoint's ingress policy?

ysksuzuki replied:
As far as I know, we apply the policy as follows.

- Pass to the stack from bpf_overlay(from-overlay), and the kernel stack forwards the packets to cilium_host(bpf_host from-host)
- [ipv[4,6]_local_delivery](https://github.com/cilium/cilium/blob/c9e89badbcc6e4989b9ffa8a38c244e80d4b102e/bpf/lib/l3.h#L60) on bpf_host handles packets to local endpoints. 
  - In the case enable-endpoint-routes=true, call redirect_ep and call [the egress program](https://github.com/cilium/cilium/blob/c9e89badbcc6e4989b9ffa8a38c244e80d4b102e/bpf/bpf_lxc.c#L2082) loaded on the veth(to-container), and apply the policy [here](https://github.com/cilium/cilium/blob/c9e89badbcc6e4989b9ffa8a38c244e80d4b102e/bpf/bpf_lxc.c#L1552)
  - In the case enable-endpoint-routes=false, call tail_call_dynamic and apply the policy [here](https://github.com/cilium/cilium/blob/c9e89badbcc6e4989b9ffa8a38c244e80d4b102e/bpf/bpf_lxc.c#L1986)

```
# The kernel stack forwards the packets to the local endpoints to cilium_host
vagrant@k8s1:~$ ip r
default via 10.0.2.2 dev enp0s3 proto dhcp src 10.0.2.15 metric 100
10.0.0.0/24 via 10.0.0.64 dev cilium_host src 10.0.0.64
10.0.0.64 dev cilium_host scope link
vagrant@k8s1:~$ kubectl get po -o wide
NAME                        READY   STATUS    RESTARTS   AGE   IP           NODE   NOMINATED NODE   READINESS GATES
testds-tlvfm                2/2     Running   0          12m   10.0.0.119   k8s1   <none>           <none>
testds-xcx65                2/2     Running   0          12m   10.0.1.154   k8s2   <none>           <none>
vagrant@k8s1:~$ ip r get 10.0.0.119
10.0.0.119 dev cilium_host src 10.0.0.64 uid 1000
    cache

# The egress programs(to-container) are attached if endpoint route is enabled
tc:
cilium_net(50) clsact/ingress cil_to_host-cilium_net id 7196
cilium_host(51) clsact/ingress cil_to_host-cilium_host id 7176
cilium_host(51) clsact/egress cil_from_host-cilium_host id 7183
lxc_health(116) clsact/ingress cil_from_container-lxc_health id 7283
lxc_health(116) clsact/egress cil_to_container-lxc_health id 7279
lxc1151b03359c3(118) clsact/ingress cil_from_container-lxc1151b03359c3 id 7326
lxc1151b03359c3(118) clsact/egress cil_to_container-lxc1151b03359c3 id 7310
lxc148700bf712d(120) clsact/ingress cil_from_container-lxc148700bf712d id 7291
lxc148700bf712d(120) clsact/egress cil_to_container-lxc148700bf712d id 7330
```

In the generic-veth chain setup, the chained CNI is responsible for the host routing (normally the chained plugin configures the kernel FIB). The kernel stack forwards packets to local endpoints to the veth and Cilium applies the policy on the to-container program which is the same as the endpoint route is enabled case.

julianwiedmann replied:
> As far as I know, we apply the policy as follows.
> 
>     * Pass to the stack from bpf_overlay(from-overlay), and the kernel stack forwards the packets to cilium_host(bpf_host from-host)
> 
>     * [ipv[4,6]_local_delivery](https://github.com/cilium/cilium/blob/c9e89badbcc6e4989b9ffa8a38c244e80d4b102e/bpf/lib/l3.h#L60) on bpf_host handles packets to local endpoints.
>       
>       * In the case enable-endpoint-routes=true, call redirect_ep and call [the egress program](https://github.com/cilium/cilium/blob/c9e89badbcc6e4989b9ffa8a38c244e80d4b102e/bpf/bpf_lxc.c#L2082) loaded on the veth(to-container), and apply the policy [here](https://github.com/cilium/cilium/blob/c9e89badbcc6e4989b9ffa8a38c244e80d4b102e/bpf/bpf_lxc.c#L1552)
>       * In the case enable-endpoint-routes=false, call tail_call_dynamic and apply the policy [here](https://github.com/cilium/cilium/blob/c9e89badbcc6e4989b9ffa8a38c244e80d4b102e/bpf/bpf_lxc.c#L1986)
> 

:+1: thanks!

julianwiedmann reviewed (changes requested):
Thanks @ysksuzuki ! Just two tiny changes from my side, and one question on policy enforcement.

nathanjsweet reviewed (approved):
LGTM for API changes.

julianwiedmann commented on bpf/Makefile:
```diff
@@ -202,8 +202,10 @@ XDP_OPTIONS = $(LB_OPTIONS) \
 	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DENABLE_DSR_HYBRID: \
 	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DDSR_ENCAP_MODE:-DDSR_ENCAP_NONE:-DDSR_ENCAP_IPIP=2 \
 	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DDSR_ENCAP_MODE:-DDSR_ENCAP_IPIP:-DDSR_ENCAP_NONE=2 \
+	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DDSR_ENCAP_MODE:-DDSR_ENCAP_GENEVE:-DDSR_ENCAP_IPIP=2 \
```
We should also add some additional configs in `bpf/complexity-tests`. But imho it's ok to add that as follow-on.

ysksuzuki replied:
We should set `bpf/complexity-tests` to maximize the number of instructions, right? So should we add `-DDSR_ENCAP_MODE=1 -DDSR_ENCAP_GENEVE=1 -DDSR_ENCAP_IPIP=2` to all configs? Will the program size be larger in GENEVE than in IP option mode?

julianwiedmann replied:
Sorry for not getting back to this earlier.

I would have expected distinct configs (ie. one with DSR-IP-Option, and a separate one for DSR-GENEVE). Otherwise we eg. lose coverage for the critical DSR-extraction path in `bpf_xdp` and `bpf_host`.

ysksuzuki replied:
I have added the complexity test configs for DSR with Geneve. Does it meet your expectation?
https://github.com/cilium/cilium/pull/23890/commits/8e6d13ec816b2e553f89172578db8b246cd04ade

julianwiedmann replied:
looks good, thank you!

nebril reviewed (approved):
Test changes LGTM, aside from the fact that we are adding new test cases to ginkgo suite. But if they cannot be run elsewhere, it's good to have this coverage.

julianwiedmann reviewed (approved):
Looks nice, thank you @ysksuzuki for the great contribution!

joestringer commented on bpf/lib/common.h:
```diff
@@ -664,6 +668,18 @@ enum metric_dir {
+ *
+ * [1]: https://www.iana.org/assignments/nvo3/nvo3.xhtml#geneve-option-class
+ */
+#define DSR_GENEVE_OPT_CLASS	0x200
```
I've applied for a Geneve Option Class for the Cilium project via IANA. Let's hold back on merging this until we hear back.

ysksuzuki replied:
Awesome! I will update it when we hear from them

joestringer replied:
We have been allocated 0x014B: https://www.iana.org/assignments/nvo3/nvo3.xhtml

ysksuzuki replied:
Thanks! I will use it.

joestringer commented on Documentation/network/kubernetes/kubeproxy-free.rst:
```diff
@@ -506,6 +507,56 @@ enabled would look as follows:
+packets, those routers drop the packets if the traffic of packets with IP options
+exceeds the threshold. That significantly affects network performance.
+
+Cilium offers another dispatch mode, DSR with Geneve to avoid this problem.
```
```suggestion
Cilium offers another dispatch mode, DSR with Geneve, to avoid this problem.
```

joestringer commented on bpf/lib/overloadable_xdp.h:
```diff
@@ -167,12 +177,29 @@ ctx_set_encap_info(struct xdp_md *ctx __maybe_unused,
+			ctx_store_meta(ctx, CB_ENCAP_PORT, *port);
+			ctx_store_meta(ctx, CB_ENCAP_ADDR, *addr);
+		} else {
+			return DROP_DSR_GENEVE_UNSUPP_IPV6_XDP;
```
Did we consider at all attempting to let such packets through to the TC layer to forward the packets, albeit at a slower rate?

joestringer replied:
(I don't consider this as a blocking comment right now, but it would be nice to at least consider this since it is introducing an incompatibility with some features that we otherwise support elsewhere)

ysksuzuki replied:
We support IPv4 in XDP using such a fallback mechanism, but we decided not to support IPv6 since the available data_meta size gets tricky when it comes to IPv6. Julian is working on the native XDP encap, which will lift the restriction for IPv6 and achieve optimal performance.

joestringer commented on bpf/lib/overloadable_xdp.h:
```diff
@@ -167,12 +177,29 @@ ctx_set_encap_info(struct xdp_md *ctx __maybe_unused,
+	if (opt) {
+		if (!is_ipv6) {
+			__be32 *addr = opt + sizeof(struct geneve_opt_hdr);
+			__be16 *port = opt + sizeof(struct geneve_opt_hdr) + sizeof(__be32);
```
Is there a reason this is doing pointer arithmetic with a `void *` rather than using proper structures? :thinking:

ysksuzuki replied:
That's because `ctx_set_encap_info`  receives either geneve_dsr_opt4 or geneve_dsr_opt6. With TC, we can just pass it to `skb_set_tunnel_opt` helper, but we need to retrieve addr and port and store them in the data_meta when it comes to XDP.

joestringer commented on pkg/monitor/api/drop.go:
```diff
@@ -85,6 +85,7 @@ var errors = map[uint8]string{
 	187: "L3 translation from IPv4 to IPv6 failed (NAT46)",
 	188: "L3 translation from IPv6 to IPv4 failed (NAT64)",
 	189: "Authentication required",
+	190: "DSR with GENEVE not supported for IPv6 with XDP",
```
I have a mild concern here that this is too specific of a drop code. If we allocated DROP codepoints for every four features that don't work together, we would exceed the remaining number space.

The drop message already contains the information that this is IPv6. I wonder if we should just make this something like:

```suggestion
	190: "Unsupported packet protocol for DSR encapsulation",
```

At least then if we ever had other encap protocols or problems outside XDP, we could use the same drop code.

joestringer commented on pkg/option/config.go:
```diff
@@ -2495,7 +2498,8 @@ func (c *DaemonConfig) TunnelDevice() string {
 func (c *DaemonConfig) TunnelExists() bool {
-	return c.TunnelingEnabled() || c.EnableIPv4EgressGateway
+	return c.TunnelingEnabled() || c.EnableIPv4EgressGateway ||
+		(c.NodePortMode == NodePortModeDSR && c.LoadBalancerDSRDispatch == DSRDispatchGeneve)
```
Is this function intended to be used with tunneling enabled for DSR? I  don't see the IPIP dispatch case here.

At a glance, it looks like this is used for Pod MTU calculation, but this shouldn't impact Pod MTUs right?

joestringer replied:
(At a glance this code looks correct, I'm just not sure if the code calling this function is accounting for this case)

ysksuzuki replied:
This function is intended to be used with or without tunneling enabled for DSR with Geneve. We send packets from intermediate nodes to backend pods through the tunnel even with the direct routing mode. So it affects Pod MTU partially. 

As far as I know, Cilium calculates the MTU considering the tunnel overhead(EthernetMTU - TunnelOverhead) if the tunnel exists, and adds route entries for pod CIDR.
```
root@kind-worker:/# ip r
default via 172.18.0.1 dev eth0
10.0.0.0/24 via 10.0.0.211 dev cilium_host src 10.0.0.211
10.0.0.211 dev cilium_host scope link
10.0.1.0/24 via 10.0.0.211 dev cilium_host src 10.0.0.211 mtu 1450
10.0.2.0/24 via 10.0.0.211 dev cilium_host src 10.0.0.211 mtu 1450
10.0.3.0/24 via 10.0.0.211 dev cilium_host src 10.0.0.211 mtu 1450
172.18.0.0/16 dev eth0 proto kernel scope link src 172.18.0.3
```

joestringer replied:
Hmm true broadly. I guess the key question here is whether (a) we think Pods should be initiating connections to remote NodePorts for East-West in-cluster forwarding, and (b) whether we think that DSR should apply to that traffic.

In the vast majority of cases, (a) should be solvable directly with ClusterIP services, which would already provide the properties of DSR without the complicated implementation. So then we could simplify the implementation by just ignoring (b) and not applying DSR to this traffic. If that's the case, then we don't need to adjust MTU for Pods based on the DSR feature.

I'm not sure I see the use case exactly for supporting DSR between Pods within the cluster, and the downside is that we end up reducing MTU for all Pods in the cluster in this mode to support this use case.

(At a high level I'm not overly concerned either way, but as long as we're looking at this area of the code, we may as well come to some sort of agreement about this and save someone the trouble of going back through this in 2+ years time when they wonder why MTU is artifically dropped in this configuration mode)

ysksuzuki replied:
We've already taken the encap overhead into account for the DSR traffic in the datapath logic. So we might not need to adjust the MTU for the node route.
https://github.com/cilium/cilium/pull/23890/files#diff-304969c111195cf4253b336fc67c9891f17b91d49738a0cc334333ac11e84b9eR278-R281

The DSR logic uses the device MTU (without overhead reduction) for the dsr_is_too_big check.
https://github.com/cilium/cilium/pull/23890/files#diff-304969c111195cf4253b336fc67c9891f17b91d49738a0cc334333ac11e84b9eR278-R281

ysksuzuki replied:
Dropped this change and left a comment. PTAL

joestringer replied:
LGTM, thanks!

julianwiedmann replied:
We'll need to adjust the naming of this helper at some point in the future ... this has now turned from `TunnelExists()` into `TunnelReducesPodMTU()`.

joestringer commented on pkg/option/config.go:
```diff
@@ -3087,8 +3091,14 @@ func (c *DaemonConfig) Populate(vp *viper.Viper) {
+			if (c.EnableNodePort || c.KubeProxyReplacement == KubeProxyReplacementStrict) &&
+				c.NodePortMode == NodePortModeDSR &&
+				c.LoadBalancerDSRDispatch == DSRDispatchGeneve {
+				c.TunnelPort = defaults.TunnelPortGeneve
```
Does this mean that egress gateway is incompatible with this new mode? Do we have checks to validate this during startup?

joestringer replied:
(I didn't see such checks, but if there are checks for egress gateway vs. DSR geneve dispatch, feel free to point them out)

ysksuzuki replied:
EgressGW is compatible with DSR Geneve dispatch. The reason for this change is that we need to set up `cilium_geneve` with port 6081 in conjunction with the direct routing, EgressGW, and DSR Geneve dispatch. Otherwise, Cilium set up `cilium_geneve` with port 8472 (vxlan port) without this change.

I have added an e2e test which checks the compatibility.
https://github.com/cilium/cilium/pull/23890/files#diff-afcd34222934f6f76dea34fd140bc589c4ef8325ba66c82f3458e1112b920622R487-R503

joestringer replied:
Doesn't this mean that the egress gateway traffic goes over port 6081 then?

If we are intending to support both features at once over two different protocols, then I don't see how we can have a single `c.TunnelPort` configuration option to support both.

ysksuzuki replied:
Yes, the egress gateway traffic goes over `cilium_geneve` port 6081 as well. 

> If we are intending to support both features at once over two different protocols, then I don't see how we can have a single c.TunnelPort configuration option to support both.

Do you mean Geneve and vxlan by different protocols? We only use Geneve for both EgressGW and DSR, and it's working.

joestringer replied:
Oh, I didn't realize that, I assumed that egress gateway just uses VXLAN. The previous code unconditionally set the port to the VXLAN port. Have we been using the wrong port the whole time?

ysksuzuki replied:
Previously the egress gateway always uses VXLAN in the direct routing, but with the DSR Geneve, we only use `cilium_geneve` because we can't use both `cilium_vxlan` and `cilium_geneve` in the current implementation. That's why this change is needed.
https://github.com/cilium/cilium/pull/23890/files#diff-3bd10cddebdc5669fc7a472e02c78702a9bc59fee30d502d1cc87a84a58439d8R349-R354

joestringer commented on bpf/lib/common.h:
```diff
@@ -664,6 +668,18 @@ enum metric_dir {
+#define DSR_GENEVE_OPT_CLASS	0x200
+#define DSR_GENEVE_OPT_TYPE	0x20
+#define DSR_IPV4_GENEVE_OPT_LEN	2
+#define DSR_IPV6_GENEVE_OPT_LEN	5
```
Can we declare these using the `sizeof` with target structures , then applying the multiplication to get the correct size? This would make the math easier to confirm it's correct. (It looks correct to me)

https://www.rfc-editor.org/rfc/rfc8926.html#name-tunnel-header-fields

ysksuzuki replied:
Is it something like this?

```
(sizeof(geneve_dsr_opt4) - sizeof(geneve_opt_hdr hdr)) / 4
(sizeof(geneve_dsr_opt6) - sizeof(geneve_opt_hdr hdr)) / 4
```

joestringer commented on bpf/lib/common.h:
```diff
@@ -664,6 +668,18 @@ enum metric_dir {
+ * [1]: https://www.iana.org/assignments/nvo3/nvo3.xhtml#geneve-option-class
+ */
+#define DSR_GENEVE_OPT_CLASS	0x200
+#define DSR_GENEVE_OPT_TYPE	0x20
```
Is there any particular pattern to this option type? `0x20` seems like an odd place to start. I would have expected to start from maybe `0x1` for the first such option in this new option class (offset by the critical bit if applicable).

It's probably also worth considering whether the 0th bit in this opt type should be set, to signal that the option is `Critical` cf. https://www.rfc-editor.org/rfc/rfc8926.html#name-tunnel-options . Given that the option contains the original service address for replying to the connection, and the remote node won't have that context, I think that this option *is* critical information that the destination node cannot ignore.

ysksuzuki replied:
There's no particular reason for `0x20`. I just picked it up randomly. It makes sense to start from `0x1`.

> It's probably also worth considering whether the 0th bit in this opt type should be set, to signal that the option is Critical

Thank you for the suggestion! I will look into that

joestringer reviewed (changes requested):
Thanks for the submission! I've reached out to IANA for a Geneve option class number allocation.

I didn't look too deeply at the datapath details for forwarding, but I have a little background with Geneve so I dropped some comments there.

joestringer commented on pkg/datapath/loader/base.go:
```diff
@@ -346,6 +346,13 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
 		args[initArgTunnelMode] = option.TunnelVXLAN
 	}
 
+	if option.Config.Tunnel == option.TunnelDisabled &&
+		option.Config.EnableNodePort &&
+		option.Config.NodePortMode == option.NodePortModeDSR &&
+		option.Config.LoadBalancerDSRDispatch == option.DSRDispatchGeneve {
+		args[initArgTunnelMode] = option.TunnelGeneve
```
What if the tunnel mode is set to `vxlan`? Seems like this would stay as "vxlan" and geneve dispatch wouldn't work?

(To be clear, I'd rather catch this way earlier during the pkg/option config parsing, not here)

ysksuzuki replied:
VXLAN with DSR doesn't work. We check it here.
https://github.com/cilium/cilium/pull/23890/files#diff-c1e209552c5426ba2f2e01f043275558eb9f8ea65553210d2f9808de17b59d1eR188-R199

joestringer commented on bpf/lib/common.h:
```diff
@@ -672,6 +676,20 @@ enum metric_dir {
+ * [1]: https://www.iana.org/assignments/nvo3/nvo3.xhtml#geneve-option-class
+ */
+#define DSR_GENEVE_OPT_CLASS	0x200
+#define DSR_GENEVE_OPT_TYPE	0x1
```
Were you able to use wireshark or something to confirm whether this is the "Critical bit" with opt type 0, or whether this is opt type 1 with no critical bit?

ysksuzuki replied:
It's opt type 1 with no critical bit. I will add critical bit next

ysksuzuki replied:
Is it OK to work on setting Critical bit as a follow-up? I think I need to find out how to set Critical bit in the Geneve header

https://www.rfc-editor.org/rfc/rfc8926.html#name-tunnel-header-fields

> Critical options present. One or more options has the critical bit set (see Section 3.5). If this bit is set, then tunnel endpoints MUST parse the options list to interpret any critical options. On tunnel endpoints where option parsing is not supported, the packet MUST be dropped on the basis of the 'C' bit in the base header. If the bit is not set, tunnel endpoints MAY strip all options using 'Opt Len' and forward the decapsulated packet. Transit devices MUST NOT drop packets on the basis of this bit.

ysksuzuki replied:
I need to set `bpf_tunnel_key.tunnel_flags`, perhaps
https://github.com/torvalds/linux/blob/1e760fa3596e8c7f08412712c168288b79670d78/drivers/net/geneve.c#L747

ysksuzuki replied:
As far as I looked into this, we can't set `TUNNEL_CRIT_OPT` to `ip_tunnel_info.key.tun_flags` using `skb_set_tunnel_key`
https://github.com/torvalds/linux/blob/1e760fa3596e8c7f08412712c168288b79670d78/net/core/filter.c#L4643-L4712

According to the Open vSwitch implementation, it sets  `TUNNEL_CRIT_OPT` to `ip_tunnel_info.key.tun_flags` if one or more options have the critical bit set
https://github.com/torvalds/linux/blob/1e760fa3596e8c7f08412712c168288b79670d78/net/openvswitch/flow_netlink.c#L2722-L2748

But there's no corresponding implementation before building a header here
https://github.com/torvalds/linux/blob/1e760fa3596e8c7f08412712c168288b79670d78/drivers/net/geneve.c#L740-L755

IMHO, we can't support Critical Bit for now. What do you think?

(We can set Ciritical Bit to the option type, but can't set the Critical bit on the Geneve header. It's not following the RFC spec.)

ysksuzuki replied:
Critical Option present: False
0x81 = 10000001 (Critical Bit + type 1)

<img width="810" alt="geneve" src="https://user-images.githubusercontent.com/5572688/227766239-92b3d7e5-e443-4836-9806-5334ea6ba684.png">

joestringer replied:
Ah I see. Adding to the option is fine for now. It looks like there's a "good first issue" for the kernel API here to either let programs provide the critical bit as a flag, or to scan the options and set the critical bit before tx. I would encourage you to follow up with the kernel folks on this if you can.

ysksuzuki replied:
>  I would encourage you to follow up with the kernel folks on this if you can.

Sure. I haven't sent a patch for the kernel API, but I will try.

qmonnet commented on Documentation/network/kubernetes/kubeproxy-free.rst:
```diff
@@ -506,6 +507,56 @@ enabled would look as follows:
+Direct Server Return (DSR) with Geneve
+**************************************
+By default, Cilium with DSR mode encodes the service IP/port in a Cilium-specific
+IPv4 option or IPv6 Destination Option extension so that the backends make aware of
```
```suggestion
IPv4 option or IPv6 Destination Option extension so that the backends are aware of
```

qmonnet commented on Documentation/network/kubernetes/kubeproxy-free.rst:
```diff
@@ -506,6 +507,56 @@ enabled would look as follows:
+However, some data center routers pass packets with unknown IP options to software
+processing called "Layer 2 slow path". As well as taking speeds from millions of
+packets, those routers drop the packets if the traffic of packets with IP options
+exceeds the threshold. That significantly affects network performance.
```
I find this paragraph hard to read. In particular, I'm not sure what you mean with “As well as taking speeds from millions of packets”. Do you mean these routers are able to process millions of packets per second? Is this relevant in the explanation?

Some suggestions:

```suggestion
However, some data center routers pass packets with unknown IP options to software
processing called "Layer 2 slow path".
Those routers drop the packets if the amount of packets with IP options
exceeds a given threshold, which may significantly affect network performance.
```

ysksuzuki replied:
I just tried to express that the router's soft process is slow, but yeah it's confusing. I'll remove it.

qmonnet commented on Documentation/network/kubernetes/kubeproxy-free.rst:
```diff
@@ -506,6 +507,56 @@ enabled would look as follows:
+Geneve dispatch enabled would look as follows:
+
+Note, IPv6 support for DSR with Geneve is currently limited. It works only with TC,
+not with XDP.
```
Please swap those two paragraphs

qmonnet reviewed (changes requested):
Thanks for this work! And apologies for the late review.

I focused mostly on the doc changes, although I glanced briefly at the rest as well and it looks good. Please find a few comments below, but it's only minor things.

qmonnet reviewed (approved):
All good from my side, thank you!

joestringer commented on bpf/lib/encap.h:
```diff
@@ -295,5 +295,76 @@ encap_and_redirect_netdev(struct __ctx_buff *ctx, struct tunnel_key *k,
 
+#if defined(ENABLE_DSR) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE
+static __always_inline int
+__encap_with_nodeid_opt(struct __ctx_buff *ctx, __u32 tunnel_endpoint,
```
IMO this code should not be referring to `nodeid` or `node_id` now that we have a concept called node id. My recommendation for this PR would be replacing existing references to `nodeid`/`node_id` with something like `tunnel_peer`, and to rename these functions as `*encap_with_dsr_opt(...)`:

```suggestion
__encap_with_dsr_opt(struct __ctx_buff *ctx, __u32 tunnel_endpoint,
```

cc @julianwiedmann @pchaigno it looks like we're starting to mix up different concepts in code like this, partly related to the way that `ctx_set_encap_info()` names the *tunnel peer* parameter (which is a destination IPv4 address) as `node_id`. I think that parameter should just be renamed in existing code.

I realize that this PR is just following the existing patterns and the nodeid naming in this code predates Paul's recent work, but that's all the more reason to be vigilant about naming in order to avoid confusing two different concepts by using the same name for both.

joestringer replied:
cf. Commit 814d3c797589 ("ipcache: Populate ipcache with node IDs")

joestringer replied:
(I would be OK with accepting this code as-is then just following up separately to cover the variable/function rename in various places across the codebase)

julianwiedmann replied:
Yes, the naming of the encap helpers hasn't aged well. And there's more churn coming (eg. adjustments needed for in-XDP encap). So let's tackle the rename separately.

joestringer commented on bpf/lib/encap.h:
```diff
@@ -295,5 +295,76 @@ encap_and_redirect_netdev(struct __ctx_buff *ctx, struct tunnel_key *k,
 }
 #endif /* TUNNEL_MODE */
 
+#if defined(ENABLE_DSR) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE
```
This function seems to be almost exactly duplicate with the existing `__encap_with_nodeid()`... Could we combine the two, but just paramaterize the `opt` parameters for the `ctx_set_encap_info()` call?

julianwiedmann replied:
I had actually played with such a variant when hacking up the in-XDP support. The verifier didn't like it. I'll have another look as part of that XDP PR, but for now I would prefer if we stick to a separate `with-option` variant.

joestringer reviewed (approved):
LGTM thanks. I noted a couple of minor cleanups that would be good to follow up in a subsequent PR below.

I would also prefer to see non-ginkgo tests for the functionality going forward since we're trying to get away from ginkgo-based testing, however I'm quite happy to see *some* tests rather than no tests. Maybe we can work together on figuring out a better way to test this functionality in future.

julianwiedmann commented on bpf/lib/common.h:
```diff
@@ -672,6 +676,20 @@ enum metric_dir {
+ * [1]: https://www.iana.org/assignments/nvo3/nvo3.xhtml#geneve-option-class
+ */
+#define DSR_GENEVE_OPT_CLASS	0x014B
+#define DSR_GENEVE_OPT_TYPE	0x81
```
In case you need to push one last time, maybe make this

```
#define GENEVE_OPT_TYPE_CRIT 0x80
...
#define DSR_GENEVE_OPT_TYPE (GENEVE_OPT_TYPE_CRIT | 0x01)
```

ysksuzuki replied:
Done

ysksuzuki committed:
bpf,test: Add complexity-test configs for DSR with Geneve

Signed-off-by: Yusuke Suzuki <yusuke-suzuki@cybozu.co.jp>

ldelossa reviewed (approved):
LGTM, nice work.
